tools/moduleConverter.py

```diff
@@ -51,39 +42,59 @@ def scrapeAllModules(str):
   return x
 
 S = set()
+Map = {}
 
 for mod in obj:
   newMod = {}
 
   moduleCode = mod["ModuleCode"]
   S.add(moduleCode)
-  if filterModule(moduleCode):
-    newMod["code"] = moduleCode
-    newMod["name"] = mod["ModuleTitle"]
-    newMod["creditCount"] = float(mod["ModuleCredit"])
-    newMod["description"] = mod.get("ModuleDescription", "")
-    newMod["preclusions"] = scrapeAllModules(mod.get("Preclusion", ""))
-    newMod["prerequisites"] = scrapeAllModules(mod.get("Prerequisite", ""))
-    newObj.append(newMod)
-    count += 1
-
+  newMod["code"] = moduleCode
+  newMod["name"] = mod["ModuleTitle"]
+  newMod["creditCount"] = float(mod["ModuleCredit"])
+  newMod["description"] = mod.get("ModuleDescription", "")
+  newMod["preclusions"] = scrapeAllModules(mod.get("Preclusion", ""))
+  newMod["prerequisites"] = scrapeAllModules(mod.get("Prerequisite", ""))
+  newObj.append(newMod)
+
+  Map[moduleCode] = newMod
+
+# Remove module codes that do not have its module info entry
+def deleteModules(newObj):
+  res = 0
+  for module in newObj:
+    prereq = module["prerequisites"]
+    for m in prereq:
+      if m not in S:
+        res += 1
+        prereq.remove(m)
+
+    preclu = module["preclusions"]
+    for m in preclu:
+      if m not in S:
+        res += 1
+        preclu.remove(m)
+
+  return res
+
+while deleteModules(newObj) > 0:
+  pass
+
+# Ensure backward preclusion relationship
 for module in newObj:
-  prereq = module["prerequisites"]
-  for m in prereq:
-    if m not in S:
-      prereq.remove(m)
-
   preclu = module["preclusions"]
-  for m in preclu:
-    if m not in S:
-      preclu.remove(m)
+  for moduleCode in preclu:
+    m = Map[moduleCode]
+    if moduleCode not in m["preclusions"] and moduleCode != module["code"]:
+      m["preclusions"].append(module["code"])
 
-print(count)
+print(len(newObj))
 
-output = os.path.join(CURRENT_DIR, "..", "src", "main", "resources", "data", "moduleInfo.json")
+# Special exceptions for certain modules
+Map["CS1231"]["prerequisites"] = [] # Remove all prereq for CS1231
```

Comment:
Other notable one is MA1521. Can add that too

Reply:
Can you list them down for me here?

Reply:
Done.